design-documents/frontend/advanced-js-bundling/advanced-js-bundling.md

```diff
@@ -0,0 +1,82 @@
+
+JavaScript module bundling is not a new idea, and Magento is far from the first piece of software to implement it. There are many excellent off-the-shelf tools designed to solve this problem:
+
+-   [Browserify](http://browserify.org/)
+-   [webpack](https://webpack.js.org/)
+-   [Parcel](https://parceljs.org/)
+
+The problem, though, is that these tools were designed for applications where the entirety of the graph can be discovered just through static analysis of JavaScript source files.
+
+This is not the case for Magento 2. Although Magento does use a proper module system where each module can express its dependencies (RequireJS), it is impossible to statically analyze which entry points will be used for an individual page.
+
+This is because RequireJS supports fetching/loading of arbitrary modules at runtime, through the global variables require and define. Unfortunately, it has become a common practice (and in some places, a documented feature) in Magento to rely on heavily dynamic use of require/define.
+
+Sources of Dynamic Module Entries in Magento 2
+
+-   [`x-magento-init`/`mage-init`](https://devdocs.magento.com/guides/v2.3/javascript-dev-guide/javascript/js_init.html)
+-   [UI Components](https://devdocs.magento.com/guides/v2.3/ui_comp_guide/bk-ui_comps.html) (`template` property)
+-   [Inline calls](https://github.com/magento/magento2/blob/07e57ac249e02b721613172c4c3c8e24a3bdd746/app/code/Magento/Sales/view/adminhtml/templates/order/create/form/address.phtml#L32)
+-   [Mixins](https://devdocs.magento.com/guides/v2.3/javascript-dev-guide/javascript/js_mixins.html)
+
+Also our [existing JS bundling](https://devdocs.magento.com/guides/v2.3/frontend-dev-guide/themes/js-bundling.html) merge *all* files into bundles which brings bigger size of static content (2x time bigger)
+
+## Problems
+
+We have 180 HTTP requests to fetch all needed JS code just for empty Home Page and overall JS size is 2.2 MB. According to google - mobile user with slow connection can have up to 8sec main-thread work time, bad pagespeed insight score, and over 21sec time to interactive.
+
+Better JS bundling approach can reduce package size by several times. This is one of the most valuable enhancement can be done from storefront side performance perspective. According to our measurements, advanced bundling can improve overall performance score, main-thread work time and TTI up to 2-3x time.
+
+## Solution
+
+I propose to introduce 'Bundlegento' as separate tool for parsing Magento store pages and generate page specific bundles.
+Currently we have prototype written in NodeJS that parse HTML pages and can generate needed bundles for Magento store. Since that this is not version specific solution, it will support all Magento released versions
+
+Also we will need to add needed functionality inside Magento that will load needed JS bundle according to layout handlers was used. JS bundle files will be added under git track and stored in theme specific location
+
```

Comment:
I am not sure if there is any benefit in this approach at it is generally a bad practice to commit any automatically generated files as this may introduce a lot of noise in commits and merge requests.

Do I assume correctly that this is done to avoid the need of running this step during a build step of deploy on a CI? If so, then maybe storing only the list of modules in every group would be enough? This would even open up a potential possibility to fiddle with those settings like adding extra modules not caught by the tool or excluding any huge ones to separate files.